pages/DSNP/UserData.md

```diff
@@ -0,0 +1,85 @@
+  * If the data type requires encryption,
+    * Retrieve the user's active (most recently announced) `keyAgreement` public key, U<sub>public</sub>.
+    If no key exists, one should be created and published as an Announcement before invoking the Operation.
+    * Create a sealed box, as in the [libsodium](https://doc.libsodium.org/public-key_cryptography/sealed_boxes) function `crypto_box_seal`, using U<sub>public</sub>.
+  * Include the previous `etag` value for the chunk. If the chunk is new, `etag` should be `null`.
+  If any chunks are to be deleted, they should be included in the input as a chunk object with the existing `etag` and a `null` value for the data field.
+
+If the Operation is invoked successfully it MUST (synchronously) return a new set of `etag` values for each data type replaced, corresponding to the updated state of the data.
+Applications should not interpret this response as an indication that the operation was completed and a state change record emitted, as this typically occurs asynchronously.
+However, this strategy allows applications to make the optimistic assumption that, in due course, the DSNP system will reflect the intended changes, without needing to wait for asynchronous confirmation.
+
+If, on the other hand, an invocation of Replace User Data is rejected, the application should fetch the most current version of the relevant data with the Get User Data Operation, reapply any intended changes, and retry the operation.
+
+The Replace User Data Operation MUST generate a [User Data Replaced Record](Records.md#user-data-replaced) containing the DSNP User Id and the set of updated user data types (but not the data itself).
+If the implementation detects that no change has occurred, it SHOULD omit the relevant unchanged data types from the state change record.
+
+# Get User Data Operation
+
+The Get User Data Operation takes the following parameters:
+* The DSNP User Id of the user who controls the data
+  * Note: While _writing_ user data is reserved for the user and any delegates, any authenticated principal can read any user's data (though it may be encrypted).
+* The user data types (by system name) that should be retrieved.
+
+The operation returns a mapping of user data type to data chunks and entity tags (the same general structure as the input data for Replace User Data), for each requested data type.
+If no chunks for a requested data type exist, an implementation MAY omit that data type from the response.
+
+To transform the public data from the output to Avro binary records, a consumer should apply the following algorithm to each data type included:
+* Determine the relevant encryption algorithm, compression codec, and object schema from the user data type and version noted.
+* For each chunk,
+  * If encryption is indicated, decrypt the chunk data using the user's secret key as in the [libsodium](https://doc.libsodium.org/public-key_cryptography/sealed_boxes) function `crypto_box_seal_open`.
```

Comment:
do we need to store any metadata along with encryption? same with Compression or since we are hardcoding this is protocol layer we are fine?

Is there any need to specify protocol version for any of these types?

Reply:
Data type and version should be enough to perform a lookup on what compression/encryption is needed. But to that point I will make it clear that version is required to be noted for each data type.